File: dashboard/src/api/executionApproval.ts
import { getAPI, postAPI } from "./client";

// Final Prompt Approval Gate — client for the per-dispatch WYSIWYG approval
// lifecycle. The operator reviews (and optionally edits) the exact provider-ready
// prompt, then approves; only an APPROVED snapshot whose execution envelope
// matches the dispatch authorises generation. See agent/api/execution_approval.py.

export type ApprovalState =
  | "REVIEW_REQUIRED"
  | "EDITED"
  | "APPROVED"
  | "INVALIDATED"
  | "DISPATCHED";

export interface ExecutionApprovalSnapshot {
  snapshot_id: string;
  approval_state: ApprovalState;
  surface: string;
  logical_mode: string;
  final_prompt_text: string;
  prompt_sha256: string;
  execution_envelope_sha256: string;
  scan_clean: number;
  scan_json?: string | null;
  approved_by?: string | null;
  approved_at?: string | null;
  invalidation_reason?: string | null;
  manifest_id?: string | null;
  manifest_item_key?: string | null;
}

// The provider-ready execution envelope frozen for review. It MUST mirror exactly
// what the surface will dispatch (same prompt + provider-affecting settings +
// resolved asset ids), so the dispatch boundary's envelope hash matches.
export interface ReviewEnvelope {
  surface: string;
  logical_mode: string;
  final_prompt_text: string;
  product_id?: string | null;
  source_mode?: string | null;
  model?: string | null;
  aspect?: string | null;
  duration_s?: number | null;
  count?: number | null;
  image_model?: string | null;
  asset_media_ids?: string[] | null;
  created_by?: string | null;
}

export function createReviewSnapshot(
  envelope: ReviewEnvelope,
): Promise<ExecutionApprovalSnapshot> {
  return postAPI<ExecutionApprovalSnapshot>("/api/execution-approval/review", envelope);
}

export function editSnapshot(
  snapshotId: string,
  editedPromptText: string,
  editorId?: string,
): Promise<ExecutionApprovalSnapshot> {
  return postAPI<ExecutionApprovalSnapshot>(
    `/api/execution-approval/${encodeURIComponent(snapshotId)}/edit`,
    { edited_prompt_text: editedPromptText, editor_id: editorId ?? null },
  );
}

export function approveSnapshot(
  snapshotId: string,
  approvedBy: string,
): Promise<ExecutionApprovalSnapshot> {
  return postAPI<ExecutionApprovalSnapshot>(
    `/api/execution-approval/${encodeURIComponent(snapshotId)}/approve`,
    { approved_by: approvedBy },
  );
}

export function invalidateSnapshot(
  snapshotId: string,
  reason: string,
): Promise<ExecutionApprovalSnapshot> {
  return postAPI<ExecutionApprovalSnapshot>(
    `/api/execution-approval/${encodeURIComponent(snapshotId)}/invalidate`,
    { reason },
  );
}

// Server-side prepare: the backend compiles the FINAL provider-ready prompt
// (product-truth grounding for IMG happens BEFORE review — never after approval)
// and freezes a REVIEW_REQUIRED snapshot. The modal shows the returned grounded
// final_prompt_text; the dispatch recomputes the SAME identity, so the approved
// prompt is dispatched EXACT. Prefer this over createReviewSnapshot for any
// surface whose prompt is grounded/compiled server-side (IMG).
export interface PrepareDispatchRequest {
  surface: string;
  logical_mode: string;
  prompt: string;
  product_id?: string | null;
  source_mode?: string | null;
  model?: string | null;
  aspect?: string | null;
  duration_s?: number | null;
  count?: number | null;
  image_model?: string | null;
  asset_media_ids?: string[] | null;
  visual_lane_id?: string | null;
  reference_pack_id?: string | null;
  creative_mode?: string | null;
  created_by?: string | null;
}

export function prepareDispatch(
  req: PrepareDispatchRequest,
): Promise<ExecutionApprovalSnapshot> {
  return postAPI<ExecutionApprovalSnapshot>("/api/execution-approval/prepare", req);
}

// --------------------------------------------------------------------------- //
// Approved Generation Manifest (multi-operation explicit approval)
// --------------------------------------------------------------------------- //

export type ManifestState = "REVIEW_REQUIRED" | "APPROVED" | "INVALIDATED";

export interface ApprovalManifest {
  manifest_id: string;
  surface: string;
  state: ManifestState;
  item_count: number;
  run_ref?: string | null;
  product_id?: string | null;
  approved_by?: string | null;
  approved_at?: string | null;
  invalidation_reason?: string | null;
  items: ExecutionApprovalSnapshot[];
}

export function getManifest(manifestId: string): Promise<ApprovalManifest> {
  return getAPI<ApprovalManifest>(
    `/api/execution-approval/manifest/${encodeURIComponent(manifestId)}`,
  );
}

export function approveManifest(
  manifestId: string,
  approvedBy: string,
): Promise<ApprovalManifest> {
  return postAPI<ApprovalManifest>(
    `/api/execution-approval/manifest/${encodeURIComponent(manifestId)}/approve`,
    { approved_by: approvedBy },
  );
}

export function invalidateManifest(
  manifestId: string,
  reason: string,
): Promise<ApprovalManifest> {
  return postAPI<ApprovalManifest>(
    `/api/execution-approval/manifest/${encodeURIComponent(manifestId)}/invalidate`,
    { reason },
  );
}

export function editManifestItem(
  manifestId: string,
  snapshotId: string,
  editedPromptText: string,
  editorId?: string,
): Promise<ApprovalManifest> {
  return postAPI<ApprovalManifest>(
    `/api/execution-approval/manifest/${encodeURIComponent(manifestId)}` +
      `/item/${encodeURIComponent(snapshotId)}/edit`,
    { edited_prompt_text: editedPromptText, editor_id: editorId ?? null },
  );
}

// Materialise the montage run's per-scene FINAL prompts into a review manifest.
export function materializeMontageManifest(runId: string): Promise<ApprovalManifest> {
  return postAPI<ApprovalManifest>(
    `/api/montage/runs/${encodeURIComponent(runId)}/materialize-approval-manifest`,
    {},
  );
}

File: agent/api/execution_approval.py
"""HTTP surface for the Final Prompt Approval Gate.

Drives the per-dispatch review -> (edit) -> approve -> (invalidate) lifecycle for
the active generation surfaces. The provider-ready execution envelope is frozen
here at review time; the dispatch boundary (make_video.start_generate and the
other credit-bearing chokes) verifies against it. Provider-free — no route in
this module ever calls a generation provider or spends a credit.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field

from agent.services import execution_approval_service as eas


router = APIRouter(prefix="/execution-approval", tags=["execution-approval"])


class ReviewSnapshotRequest(BaseModel):
    surface: str
    logical_mode: str
    final_prompt_text: str = Field(..., min_length=1)
    product_id: str | None = None
    source_mode: str | None = None
    model: str | None = None
    aspect: str | None = None
    duration_s: int | None = None
    count: int | None = None
    image_model: str | None = None
    asset_media_ids: list[str] | None = None
    review_session_id: str | None = None
    created_by: str | None = None


class EditSnapshotRequest(BaseModel):
    edited_prompt_text: str = Field(..., min_length=1)
    editor_id: str | None = None


class ApproveSnapshotRequest(BaseModel):
    approved_by: str = Field(..., min_length=1)


class InvalidateSnapshotRequest(BaseModel):
    reason: str = Field(..., min_length=1)


def _raise(exc: eas.ExecutionApprovalError):
    raise HTTPException(
        status_code=exc.status_code,
        detail={"error": exc.code, "message": exc.message, "details": exc.details},
    ) from exc


@router.post("/review")
async def create_review(req: ReviewSnapshotRequest) -> dict[str, Any]:
    return await eas.create_review_snapshot(
        surface=req.surface,
        logical_mode=req.logical_mode,
        final_prompt_text=req.final_prompt_text,
        product_id=req.product_id,
        source_mode=req.source_mode,
        model=req.model,
        aspect=req.aspect,
        duration_s=req.duration_s,
        count=req.count,
        image_model=req.image_model,
        asset_media_ids=req.asset_media_ids,
        review_session_id=req.review_session_id,
        created_by=req.created_by,
    )


@router.get("/{snapshot_id}")
async def get_snapshot(snapshot_id: str) -> dict[str, Any]:
    snap = await eas.get_snapshot(snapshot_id)
    if snap is None:
        raise HTTPException(status_code=404, detail={"error": "SNAPSHOT_NOT_FOUND"})
    return snap


@router.get("")
async def list_snapshots(
    product_id: str | None = Query(default=None),
    surface: str | None = Query(default=None),
    limit: int = Query(default=50, ge=1, le=200),
) -> dict[str, Any]:
    from agent.db import execution_approval_crud as _crud

    items = await _crud.list_snapshots(product_id=product_id, surface=surface, limit=limit)
    return {"items": items, "count": len(items)}


@router.post("/{snapshot_id}/edit")
async def edit_snapshot(snapshot_id: str, req: EditSnapshotRequest) -> dict[str, Any]:
    try:
        return await eas.apply_edit(
            snapshot_id, edited_prompt_text=req.edited_prompt_text, editor_id=req.editor_id,
        )
    except eas.ExecutionApprovalError as exc:
        _raise(exc)


@router.post("/{snapshot_id}/approve")
async def approve_snapshot(snapshot_id: str, req: ApproveSnapshotRequest) -> dict[str, Any]:
    try:
        return await eas.approve_snapshot(snapshot_id, approved_by=req.approved_by)
    except eas.ExecutionApprovalError as exc:
        _raise(exc)


@router.post("/{snapshot_id}/invalidate")
async def invalidate_snapshot(snapshot_id: str, req: InvalidateSnapshotRequest) -> dict[str, Any]:
    try:
        return await eas.invalidate_snapshot(snapshot_id, reason=req.reason)
    except eas.ExecutionApprovalError as exc:
        _raise(exc)
